sdk/core/azure-core/src/main/java/com/azure/core/http/policy/AzureSasCredentialPolicy.java

```diff
@@ -0,0 +1,60 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.azure.core.http.policy;
+
+import com.azure.core.credential.AzureSasCredential;
+import com.azure.core.http.HttpPipelineCallContext;
+import com.azure.core.http.HttpPipelineNextPolicy;
+import com.azure.core.http.HttpRequest;
+import com.azure.core.http.HttpResponse;
+import reactor.core.publisher.Mono;
+
+import java.util.Objects;
+
+/**
+ * Pipeline policy that uses an {@link AzureSasCredential} to set the shared access signature for a request.
+ * <p>
+ * Requests sent with this pipeline policy are required to use {@code HTTPS}. If the request isn't using {@code HTTPS}
+ * an exception will be thrown to prevent leaking the shared access signature.
+ */
+public final class AzureSasCredentialPolicy implements HttpPipelinePolicy {
+    private final AzureSasCredential credential;
+
+    /**
+     * Creates a policy that uses the passed {@link AzureSasCredential} to append sas to query string.
+     *
+     * @param credential The {@link AzureSasCredential} containing the shared access signature to use.
+     * @throws NullPointerException If {@code credential} is {@code null}.
+     */
+    public AzureSasCredentialPolicy(AzureSasCredential credential) {
+        Objects.requireNonNull(credential, "'credential' cannot be null.");
+
+        this.credential = credential;
+    }
+
+    @Override
+    public Mono<HttpResponse> process(HttpPipelineCallContext context, HttpPipelineNextPolicy next) {
+        HttpRequest httpRequest = context.getHttpRequest();
+        if ("http".equals(httpRequest.getUrl().getProtocol())) {
+            return Mono.error(new IllegalStateException(
+                "Shared access signature credentials require HTTPS to prevent leaking the shared access signature."));
+        }
+
+        String signature = credential.getSignature();
+        if (signature.startsWith("?")) {
```

Comment:
Do we want to clean the signature on construction of the credential to reduce the number of times we need to substring?

Reply:
```
AzureSasCredential credential = AzureSasCredential("?foo");
System.out.println(credential.getSignature()); // this would print "foo"
```
I was thinking about it but I thought it would be good to make sure that what we feed into constructor is the same what getSignature() returns.

Reply:
Makes sense to me, the sub-stringing is only a minor cost anyhow.

Reply:
Stripping out the leading `?` may work for storage but will there be other services where the leading `?` would still be required?